celery_pubsub/pubsub.py

```diff
@@ -1,61 +1,37 @@
+import re
 import celery
 
-__all__ = ['publish', 'publish_now', 'subscribe', 'unsubscribe', ]
-
 
 class PubSubManager(object):
+
     def __init__(self):
-        super(PubSubManager, self).__init__()
-        self.subscribed = []
         self.jobs = {}
 
     def publish(self, topic, *args, **kwargs):
-        result = self.get_jobs(topic).delay(*args, **kwargs)
-        return result
+        group = []
+        for (regex, task) in self.jobs.values():
+            if regex.match(topic):
+                group.append(task)
+        out = celery.group(group)
+        return out
 
     def publish_now(self, topic, *args, **kwargs):
-        result = self.get_jobs(topic).apply(args=args, kwargs=kwargs)
+        # REVIEW: useless shortcut?
+        result = self.publish(topic, *args, **kwargs).apply(args=args, kwargs=kwargs)
         return result
 
     def subscribe(self, topic, task):
-        regex = self._topic_to_re(topic)
-        if (topic, regex, task) not in self.subscribed:
-            self.subscribed.append((topic, regex, task))
-            self.jobs = {}
-
-    def unsubscribe(self, topic, task):
-        found = -1
-        for idx, triplet in enumerate(self.subscribed):
-            if (triplet[0], triplet[2]) == (topic, task):
-                found = idx
-                break
-        if not found == -1:
-            del self.subscribed[found]
-            self.jobs = {}
-
-    def get_jobs(self, topic):
-        if topic not in self.jobs:
-            self._gen_jobs(topic)
-        return self.jobs[topic]
-
-    def _gen_jobs(self, topic):
-        jobs = []
-        for job in self.subscribed:
-            if job[1].match(topic):
-                jobs.append(job[2].s())
-        self.jobs[topic] = celery.group(jobs)
-
-    @staticmethod
-    def _topic_to_re(topic):
         assert isinstance(topic, str)
-        import re
         re_topic = topic.replace('.', r'\.').replace('*', r'[^.]+').replace('#', r'.+')
-        return re.compile(r'^{}$'.format(re_topic))
+        # REVIEW: what happens if TOPIC already exists?
+        self.jobs[topic] = (regex, task)
+
+    def unsubscribe(self, topic, task):
+        del self.jobs[topic]
 
 
-_pubsub_manager = None
-if _pubsub_manager is None:
-    _pubsub_manager = PubSubManager()
+# Hence the class is useless?
+_pubsub_manager = PubSubManager()
```

Comment:
If you intent the users of your project to inherit the `PubSubManager` class (for some reason) you should keep the class, otherwise, my advice is to avoid classes wherever possible. At least `class` should not be follow the hype mimetic meme.

Reply:
The class is to group the functionalities together.
Moreover, I use the Pimpl pattern ( https://www.geeksforgeeks.org/pimpl-idiom-in-c-with-examples/ ) as an approach to show exactly which contract I'll guarantee my API users. These 4 functions (below) will not move, even if the implementation details may completely change in the future.